### ISSUE TITLE
Retrain Compositional Embeddings and Tree-LSTM on M1 Corpus

### ISSUE THREAD

**Issue by marctjones** [marctjones/klareco#34]
## Problem: Training Data Mismatch

**Current models trained on**:
- Tatoeba EN-EO parallel corpus: 271K sentence pairs
- Domain: Short, conversational, user-contributed
- Vocabulary: Limited to Tatoeba roots (~10K unique roots)
- Date: April 2023 snapshot

**M1 corpus will be**:
- Wikipedia: 4.2M encyclopedic sentences
- Books: 26.8K literary sentences
- Domain: Encyclopedic, formal, technical content
- Vocabulary: 15× more data = much broader root coverage

### 1. Vocabulary Coverage Gap

Current embeddings know roots from conversational Tatoeba:
- ✅ "hundo" (dog), "kato" (cat), "manĝi" (eat)

Wikipedia has many unknown roots:
- ❌ "kvantuma" (quantum), "ribonuklea" (ribonucleic)
- ❌ "Mezopotamio" (Mesopotamia), "Adriatika" (Adriatic)
- ❌ "Bizanca" (Byzantine), "mezepoko" (middle ages)

**Result**: Unknown roots will get random/poor embeddings → poor retrieval

### 2. Domain Mismatch

- **Tatoeba**: "Mi manĝas pomon" (I eat an apple) - casual
- **Wikipedia**: "La Bizanca Imperio ekzistis de 330 ĝis 1453" - encyclopedic

Semantic similarity learned on conversational pairs may not transfer well to encyclopedic retrieval.

### 3. Embedding Quality for RAG

Current model optimized for:
- Paraphrase detection (same English → similar Esperanto)
- Conversational similarity

M1 RAG needs:
- Topical similarity (encyclopedic concepts)
- Question-answer matching
- Factual retrieval

## What Needs Retraining

**Models to retrain** (2.6M params total):
1. **Compositional Embeddings** (640K params)
   - Root embeddings: [10K × 64] → may need expansion to [15K-20K × 64]
   - Prefix/suffix embeddings: Can stay same (fixed Esperanto grammar)
   - Ending embeddings: Already deterministic

2. **Tree-LSTM Encoder** (2M params)
   - Architecture can stay same
   - Weights need retraining on new domain

3. **RAG Retriever**
   - Deterministic logic stays same
   - Just needs new trained models

**Keep as-is**:
- ✅ Model architecture (`klareco/embeddings/compositional.py`)
- ✅ Training code (`scripts/train_semantic_similarity.py`)
- ✅ Parser, deparser (deterministic, 0 params)

### Option A: Self-Supervised on M1 Corpus (Preferred)

**No external parallel data needed!**

Train on M1 corpus itself using self-supervised objectives:

**1. Masked Language Modeling (AST-level)**
- Mask random nodes in AST
- Predict masked content from context
- Learns semantic relationships within corpus

**2. Next Sentence Prediction**
- Predict if sentence B follows sentence A in article
- Uses Wikipedia article structure
- Learns topical coherence

**3. Contrastive Learning**
- Sentences from same Wikipedia article = positive pairs
- Sentences from different articles = negative pairs
- Learns semantic clustering

**Advantages**:
- ✅ No external dependencies
- ✅ Learns from actual M1 domain
- ✅ Vocabulary coverage guaranteed
- ✅ 4.2M sentences = plenty of training signal

**Training data**: 4.2M Wikipedia + 26K books = self-supervised from M1 corpus

### Option B: Generate New Similarity Pairs

**If we need explicit similarity labels**:

1. Use GPT-4 to generate Esperanto paraphrase pairs from Wikipedia sentences
2. Use sentence-transformers on Esperanto directly (multilingual models)
3. Generate Q&A pairs (Issue #24) and use for semantic training

**Disadvantages**:
- ⚠️ External dependency (GPT-4 or multilingual model)
- ⚠️ May not preserve linguistic purity
- ⚠️ More expensive

### Phase 1: Extract Training Data from M1 Corpus

**After Issue #28 (Build Enhanced Corpus)**:

```python
# scripts/generate_self_supervised_pairs.py

# Positive pairs (same article, nearby sentences)
for article in corpus:
    for i, sent_a in enumerate(article.sentences):
        # Nearby sentences (topically related)
        for sent_b in article.sentences[i+1:i+10]:
            pairs.append((sent_a, sent_b, similarity=0.8))

# Negative pairs (different articles)
for sent_a in random_sample(corpus):
    for sent_b in random_sample(corpus):
        if sent_a.article != sent_b.article:
            pairs.append((sent_a, sent_b, similarity=0.1))
```

### Phase 2: Update Vocabulary

```python
# Build new root vocabulary from M1 corpus
roots = extract_all_roots(m1_corpus)  # 15K-20K roots expected

# Initialize new embeddings
old_emb = load('models/semantic_similarity/best_model.pt')
new_emb = expand_vocabulary(old_emb, new_roots)
# Transfer learned weights for known roots
# Initialize random for new roots
```

### Phase 3: Retrain Models

**Same training loop as before**:
```bash
python scripts/train_semantic_similarity.py \
  --train-data data/m1_similarity_pairs_train.jsonl \
  --val-data data/m1_similarity_pairs_val.jsonl \
  --output-dir models/semantic_similarity_m1 \
  --epochs 50
```

**Expected duration**: 1-2 weeks on GPU

### Phase 4: Evaluate & Replace

**Evaluation**:
- Test on M1 corpus sentences (not Tatoeba)
- Measure retrieval quality (precision@k)
- Compare old vs new embeddings on Wikipedia queries

**If better**:
- Replace `models/semantic_similarity/best_model.pt`
- Update RAG retriever to use new model
- Archive old model to `models/semantic_similarity_tatoeba/` (for reference)

## Dependencies

**Blocks**:
- Issue #28: Build Enhanced Corpus (need M1 corpus first)

**Blocked by this**:
- Issue #24: Generate Q&A Pairs (should use new embeddings)
- Issue #25: Train Reasoning Core (should use new embeddings)
- Issue #30: Reranker (should use new embeddings)

## Priority

**P0** - Critical Path

This is actually a **prerequisite** for effective M1 training. We need embeddings that understand M1's vocabulary and domain.

**Recommended order**:
1. Issue #28: Build Enhanced Corpus
2. **THIS ISSUE**: Retrain embeddings on M1 corpus
3. Issue #24: Generate Q&A pairs
4. Issue #25: Train reasoning core

## Estimated Effort

- Data generation: 2-3 days
- Vocabulary expansion: 1 day
- Training: 1-2 weeks (GPU dependent)
- Evaluation: 2-3 days

**Total**: 2-3 weeks

## Open Questions

1. **Self-supervised objective**: Which works best for Esperanto encyclopedic text?
   - Masked LM? Next sentence? Contrastive? All three?

2. **Vocabulary size**: Expand to 15K, 20K, or dynamic based on M1 corpus?

3. **Transfer learning**: Start from Tatoeba weights or train from scratch?
   - Pro transfer: Faster convergence
   - Pro scratch: No conversational bias

4. **Evaluation metric**: How to measure "better" for encyclopedic retrieval?
   - Manual review of top-K retrieved sentences?
   - Synthetic Q&A pairs?

## Related Issues

- Issue #28: Build Enhanced Corpus (prerequisite)
- Issue #24: Generate Q&A Pairs (should use new embeddings)
- Issue #25: Train Reasoning Core (should use new embeddings)
- Issue #30: Reranker (should use new embeddings)
- Issue #33: Cleanup (old Tatoeba model goes to deprecated/)

**Comment by marctjones** [comment]
## ✅ Progress Update - Issues #41, #42, #43 Complete!

### Completed Tasks

**Issue #41: Root Frequency Analysis** ✅ CLOSED
- Analyzed 4.2M sentences, 78M word instances
- Found 2.1M unique roots (includes proper nouns, numbers, etc.)
- **Recommendation: 5,000 root vocabulary for 81.2% coverage**
- Top 5K roots cover vast majority of encyclopedic content
- Rare roots can use <UNK> + compositional inference

**Issue #42: Affix Classification** ✅ CLOSED  
- Created `config/affix_classification.json`
- 8 semantic prefixes + 15 semantic suffixes (need 128d embeddings)
- 4 grammatical prefixes + 13 grammatical suffixes (need 16d embeddings)
- 17 endings (one-hot encoding)

**Issue #43: MorphemeAwareEmbedding Model** ✅ CLOSED
- Implemented `klareco/embeddings/morpheme_aware.py`
- Separates semantic (128d) vs grammatical (16d) components
- Attention-based composition for semantic affixes
- **Total parameters: ~709K** (vs 3.9M for baseline!)
- All tests passing ✅

### Architecture Summary

```python
MorphemeAwareEmbedding(
    root_vocab_size=5000,      # 640K params (81.2% coverage)
    semantic_prefixes=8,        # 1K params
    semantic_suffixes=15,       # 2K params
    grammatical_prefixes=4,     # 64 params
    aspect_suffixes=13,         # 208 params
    endings=17,                 # 272 params
    composer=MultiheadAttention # 66K params
)
# Total: ~709K params (81% reduction from 3.9M!)
```

### Next: Issue #44 (Implement Training Script)

Now ready to implement the morpheme-aware training script with:
- Masked root prediction (learn root semantics)
- Affix composition (learn affix transformations)
- Grammatical feature prediction (learn syntax patterns)

Estimated: 2-3 hours implementation + 1-1.5 hours training

**Comment by marctjones** [comment]
Closing: Tree-LSTM approach abandoned in favor of compositional embeddings (Stage 1). See Training Plan v3 (#100).
